### PR TITLE
Add support for `cameraForBounds` on globe projection

### PR DESCRIPTION
**Pull request by karimnaaji**
Fixes https://github.com/mapbox/gl-js-team/issues/479

This PR adds support for `cameraForBounds` on globe and updates the method used by `cameraForBounds` to account fully for 3d perspective. A demo of the prototype to help visualize the new method is available [here](https://karim.naaji.fr/frustum-fit.html).

Screen Shot 2022-08-04 at 5 01 20 PM

Currently, `cameraForBounds` only takes into account the map `bearing` in 2d dimension. In order to use it on globe as well as extend the method to account for `pitch` appropriately, the camera frustum adjustment has to be done in using the 3d bounding box of the coordinates we are fitting.

The new method reconstructs a 3d bounding box from the perspective of the camera, as can be seen in the following video:

https://user-images.githubusercontent.com/7061573/182975631-35471466-c537-434f-975a-2ec90d9cab67.mov

It's important to note that left and right planes aren't perfectly tight with the initial bounding box; this is an overall preferable behavior as we ensure that that the final camera placement is looking at the centroid of both the reconstructed aabb in camera space and the original bounds.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [x] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Add support for cameraForBounds with globe view</changelog>`

### CODE REVIEW

**Comment by mpulkki-mapbox** [comment]
These animations are awesome and really help to understand idea behind the code! I have few questions as this is a really fascinating topic!

I see that you're finding the best fit for the camera frustum using an AABB in camera's coordinate space that's been built by encapsulating the AABB transformed from the ecef space. Why do you think it's necessary step to construct an intermediate AABB in ecef space first? Right now result of the algorithm seems to vary as a function of longitude and latitude as the bounding box is aligned in ecef space and not in camera space (your interactive demo is _really_ useful for spotting things like these!)

Taking this route you could first try to transform corners of the lat&lng bounds into world space (aka pixel space) using the globe matrix and then building the AABB. This could help with the aforementioned issue as the up vector in world space is always pointing to same direction because camera actually moves in a flat plane in mercator coordinates. In other words intermediate AABBs built this way should always be aligned with the tangent plane at the centroid location (I think).

A slightly modified approach I've been thinking about is to simplify the problem into fitting points inside a frustum. It _should_ be a fairly simple problem as camera centroid, bearing & pitch are fixed and only distance to the ground is changing. Given a set of points (corners of the lat&lng region in this case) and camera frustum planes you could find the closest intersection point between any pair of frustum planes and points and use that distance to move the camera forward.

**Comment by karimnaaji** [comment]
>These animations are awesome and really help to understand the idea behind the code! I have few questions as this is a really fascinating topic!

Thanks for giving a look!

>I see that you're finding the best fit for the camera frustum using an AABB in camera's coordinate space that's been built by encapsulating the AABB transformed from the ecef space. Why do you think it's necessary step to construct an intermediate AABB in ecef space first? Right now result of the algorithm seems to vary as a function of longitude and latitude as the bounding box is aligned in ecef space and not in camera space (your interactive demo is really useful for spotting things like these!)
>
>Taking this route you could first try to transform corners of the lat&lng bounds into world space (aka pixel space) using the globe matrix and then building the AABB. This could help with the aforementioned issue as the up vector in world space is always pointing to same direction because camera actually moves on a flat plane in mercator coordinates. In other words intermediate AABBs built this way should always be aligned with the tangent plane at the centroid location (I think).

Yes that's a good point, it's worth investigating whether we could skip an extra space transformation in order to reduce the longitude/latitude dependency by doing it directly in world space as you suggest, I'll investigate that as it might might also result in a better fit overall.

>A slightly modified approach I've been thinking about is to simplify the problem into fitting points inside a frustum. It should be a fairly simple problem as camera centroid, bearing & pitch are fixed and only distance to the ground is changing. Given a set of points (corners of the lat&lng region in this case) and camera frustum planes you could find the closest intersection point between any pair of frustum planes and points and use that distance to move the camera forward.

I investigated this, but unless I missed something, finding the closest intersection of the frustum planes aligned with a set of points may not always result in the camera being aligned or looking at the centroid, which I think is an important property of `cameraForBounds` placement API. Although it would result in tighter placement, keeping the centroid of the bounds being fit aligned with the center (as implemented in this PR) is a tradeoff.

File: src/util/primitives.js [hunk 1]
```diff
@@ -185,12 +185,26 @@ class Aabb {
         return new Aabb(min, max);
     }
 
+    clone(): Aabb {
+        return new Aabb(vec3.clone(this.min), vec3.clone(this.max));
+    }
```
Comment (mourner) [inline src/util/primitives.js:190]: Unused in the PR, should probably be removed?

File: src/ui/camera.js [hunk 2]
```diff
@@ -632,66 +756,102 @@ class Camera extends Evented {
...
+        aabb.setMin([
+            aabb.min[0] - (padL + halfScreenPadX) * scaleRatio,
+            aabb.min[1] - (padB + halfScreenPadY) * scaleRatio,
+            aabb.min[2]
+        ]);
+        aabb.setMax([
+            aabb.max[0] + (padR + halfScreenPadX) * scaleRatio,
+            aabb.max[1] + (padT + halfScreenPadY) * scaleRatio,
+            aabb.max[2]
+        ]);
```
Comment (mourner) [inline src/ui/camera.js:812]: You're basically replacing the Aabb instance completely, so you could just remove `setMin`/`setMax` methods and do `new Aabb` for simplicity.

File: src/ui/camera.js [hunk 3]
```diff
@@ -632,66 +756,102 @@ class Camera extends Evented {
...
+        const screenPadB = edgePadding.bottom || 0;
+        const screenPadT = edgePadding.top || 0;
+
+        const padL = eOptions.padding.left;
+        const padR = eOptions.padding.right;
+        const padT = eOptions.padding.top;
+        const padB = eOptions.padding.bottom;
+
+        const halfScreenPadX = ((edgePadding.left || 0) + (edgePadding.right || 0)) * 0.5;
+        const halfScreenPadY = ((edgePadding.top || 0) + (edgePadding.bottom || 0)) * 0.5;
```
Comment (mourner) [inline src/ui/camera.js:794]: There are already variables for `edgePadding.left || 0` / etc. above.

File: src/ui/camera.js [hunk 4]
```diff
@@ -632,66 +756,102 @@ class Camera extends Evented {
...
+
+        const screenPadL = edgePadding.left || 0;
+        const screenPadR = edgePadding.right || 0;
+        const screenPadB = edgePadding.bottom || 0;
+        const screenPadT = edgePadding.top || 0;
+
+        const padL = eOptions.padding.left;
+        const padR = eOptions.padding.right;
+        const padT = eOptions.padding.top;
+        const padB = eOptions.padding.bottom;
```
Comment (mourner) [inline src/ui/camera.js:791]:
```suggestion
        const {left: padL, right: padR, top: padT, bottom: padB} = eOptions.padding;
```

File: src/util/primitives.js [hunk 5]
```diff
@@ -204,6 +218,26 @@ class Aabb {
...
+        this.max = [-Infinity, -Infinity, -Infinity];
+
+        for (const p of corners) {
+            vec3.min(this.min, this.min, p);
+            vec3.max(this.max, this.max, p);
+        }
+
+        this.center = vec3.scale([], vec3.add([], this.min, this.max), 0.5);
+
+        return this;
```
Comment (mourner) [inline src/util/primitives.js:238]: If you make this method immutable, you could remove some repetitive code:

```suggestion
        return Aabb.fromPoints(corners);
```

File: src/geo/transform.js [hunk 6]
```diff
@@ -2075,11 +2109,23 @@ class Transform {
         }
     }
 
-    getCameraToCenterDistance(projection: Projection): number {
-        const t = getProjectionInterpolationT(projection, this.zoom, this.width, this.height, 1024);
+    getCameraToCenterDistance(projection: Projection, zoom: ?number): number {
+        zoom = zoom || this.zoom;
```
Comment (mourner) [inline src/geo/transform.js:2113]: Nit: it's more idiomatic to use default params for things like this (and you can use `this` there too)

File: src/ui/camera.js [hunk 7]
```diff
@@ -609,6 +617,122 @@ class Camera extends Evented {
...
+
+        ecefCoords.push(latLngToECEF(coord0.lat, coord0.lng));
+        ecefCoords.push(latLngToECEF(coord1.lat, coord0.lng));
+        ecefCoords.push(latLngToECEF(coord1.lat, coord1.lng));
+        ecefCoords.push(latLngToECEF(coord0.lat, coord1.lng));
+
+        ecefCoords.push(latLngToECEF((coord0.lat + coord1.lat) * 0.5, coord0.lng));
+        ecefCoords.push(latLngToECEF((coord0.lat + coord1.lat) * 0.5, coord1.lng));
+        ecefCoords.push(latLngToECEF(coord0.lat, (coord0.lng + coord1.lng) * 0.5));
+        ecefCoords.push(latLngToECEF(coord1.lat, (coord0.lng + coord1.lng) * 0.5));
```
Comment (mourner) [inline src/ui/camera.js:677]: Nit: array literal instead of pushes for simplicity.

File: src/ui/camera.js [hunk 8]
```diff
@@ -609,6 +617,122 @@ class Camera extends Evented {
...
+            min[0] = Math.min(min[0], vec3.dot(xAxis, p));
+            min[1] = Math.min(min[1], vec3.dot(yAxis, p));
+            min[2] = Math.min(min[2], vec3.dot(zAxis, p));
+
+            max[0] = Math.max(max[0], vec3.dot(xAxis, p));
+            max[1] = Math.max(max[1], vec3.dot(yAxis, p));
+            max[2] = Math.max(max[2], vec3.dot(zAxis, p));
+        }
+
+        const aabb = new Aabb(min, max);
```
Comment (mourner) [inline src/ui/camera.js:689]: You could do `ecefCoords.map(p => [vec3.dot(xAxis, p), vec3.dot(yAxis, p), vec3.dot(zAxis, p)])` and then use `Aabb.fromPoints` to remove some more repetitive code.

**Review by mourner — CHANGES_REQUESTED** [review]
I didn't get through all the math logic in detail, but basically I think this needs a few more passes of simplification — the code is a bit too verbose / repetitive in places, and adds quite a bit of bundle size (+1.3KB gzipped!) for the feature. Pointed out some low hanging fruits but there should be more opportunities to cut down / simplify.

File: src/ui/camera.js [hunk 9]
```diff
@@ -609,6 +617,122 @@ class Camera extends Evented {
...
+
+        const center = vec3.transformMat4([], aabb.center, aabbOrientation);
```
Comment (mpulkki-mapbox) [inline src/ui/camera.js:691]: Is there a difference between using `aabb.center` or `origin` as the center point?

**Review by mpulkki-mapbox — COMMENTED** [review]
Looks and works great! I have few minor questions in the comments

File: src/util/primitives.js (hunk 1 above)
Comment (karimnaaji) [reply src/util/primitives.js]: Yes leftover that ended up being unused, I've removed it in latest commits.

File: src/util/primitives.js [hunk 10]
```diff
@@ -185,12 +185,26 @@ class Aabb {
         return new Aabb(min, max);
     }
 
+    clone(): Aabb {
+        return new Aabb(vec3.clone(this.min), vec3.clone(this.max));
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/util/primitives.js]: Good catch, I've removed this method as it ended up being unused.

File: src/ui/camera.js [hunk 11]
```diff
@@ -609,6 +617,122 @@ class Camera extends Evented {
...
+
+        vec3.normalize(center, center);
+        vec3.scale(center, center, GLOBE_RADIUS);
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/ui/camera.js:696]: Good catch, I'll add a test and support for this corner case.

File: src/geo/transform.js [hunk 12]
```diff
@@ -1969,6 +1980,29 @@ class Transform {
...
+    zoomFromMercatorZAdjusted(z: number): number {
+        const getZoom = (zoom) => {
+            const d = this.getCameraToCenterDistance(this.projection, zoom);
+            return this.scaleZoom(d / (z * this.tileSize));
+        };
+
+        let zoom = getZoom(this.zoom);
+        let diff = Math.abs(zoom - getZoom(zoom));
+        let lastdiff;
+        while (lastdiff !== diff) {
```
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/geo/transform.js:1997]: It should converge in a few iterations, but as a safeguard it would be wiser to add a max iteration count or an epsilon comparison, as it's difficult to confirm that floating point comparison will not break this. This method will not be necessary in gl-native, as adaptive scaling to get constant globe size isn't applied yet.

**Comment by karimnaaji** [comment]
Thanks for the review and helpful simplification suggestions, I've addressed most of them and will add more tests for the corner case of requesting entries such as [-180, 0], [180, 0]. Regarding the concern of size:

>and adds quite a bit of bundle size (+1.3KB gzipped!)

it will level-out with https://github.com/mapbox/mapbox-gl-js/pull/12211 (once `fitScreenCoordinates` and `fitBounds` all go through using the method introduced here, allowing to delete some code and benefit from globe support), I've split the work in two separate PRs to make it easier to review.

**Review by SnailBones — COMMENTED** [review]
I'm seeing some strange behavior with bounds surrounding a pole. The map usually zooms much further out than I would expect. Sometimes, when starting at high zoom and pitch, zoom doesn't change at all.

https://user-images.githubusercontent.com/14878684/188241598-32dd3180-6aea-477e-81de-427726799c18.mov


I'm also noticing that the globe will spin around rather than crossing the antimeridian, this seems to be an existing issue with `map.easeTo()`.

https://user-images.githubusercontent.com/14878684/188241372-4e2020d6-1396-48c7-b141-eb2492c71920.mov

File: src/geo/projection/globe_util.js [hunk 13]
```diff
@@ -428,6 +428,15 @@ function csLatLngToECEF(cosLat: number, sinLat: number, lng: number, radius: num
     return [sx, sy, sz];
 }
 
+export function ecefToLatLng(ecef: Array<number>): LngLat {
+    const radius = Math.sqrt(ecef[0] * ecef[0] + ecef[1] * ecef[1] + ecef[2] * ecef[2]);
+
+    const lng = Math.atan2(ecef[0], ecef[2]);
+    const lat = Math.PI * 0.5 - Math.acos(-ecef[1] / radius);
```
Comment (mourner) [inline src/geo/projection/globe_util.js:435]: Nit: much nicer with destructuring in cases like this.

```suggestion
export function ecefToLatLng([x, y, z]: Array<number>): LngLat {
    const radius = Math.hypot(x, y, z);
    const lng = Math.atan2(x, z);
    const lat = Math.PI * 0.5 - Math.acos(-y / radius);
```

File: src/ui/camera.js [hunk 14]
```diff
@@ -609,6 +617,113 @@ class Camera extends Evented {
...
+        const aabbAspectRatio = aabbW / aabbH;
+        const selectXAxis = aabbAspectRatio > tr.aspect;
+
+        const min = aabb.min;
+        const max = [
+            selectXAxis ? aabb.max[0] : aabb.min[0],
+            selectXAxis ? aabb.min[1] : aabb.max[1],
+            aabb.min[2]
+        ];
+        const axisLength = vec3.length(vec3.sub([], min, max));
```
Comment (mourner) [inline src/ui/camera.js:632]: Looks like this code is mostly redundant, since `max[2] - min[2]` is always `0`, and two other components are mutually exclusive, so the vector is axis-aligned.

```suggestion
        const axisLength = selectXAxis ? aabbW : aabbH;
```

File: src/ui/camera.js [hunk 15]
```diff
@@ -609,6 +617,113 @@ class Camera extends Evented {
...
+
+        const aabbHalfExtentZ = (aabb.max[2] - aabb.min[2]) * 0.5;
+        const frustumDistance = this._minimumAABBFrustumDistance(tr, aabb);
+
+        const offsetZ = vec3.scale([], [0, 0, 1], aabbHalfExtentZ);
+        const aabbClosestPoint = vec3.add(offsetZ, center, offsetZ);
+        const offsetDistance = frustumDistance + (tr.pitch === 0 ? 0 : vec3.distance(center, aabbClosestPoint));
+
+        const globeCenter = tr.globeCenterInViewSpace;
+        const normal = vec3.sub([], center, [globeCenter[0], globeCenter[1], globeCenter[2]]);
```
Comment (mourner) [inline src/ui/camera.js:702]: Nit: `globeCenter` is only read here:

```suggestion
        const normal = vec3.sub([], center, globeCenter);
```

File: src/ui/camera.js [hunk 16]
```diff
@@ -632,66 +747,92 @@ class Camera extends Evented {
...
+
+        const zoomRef = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), eOptions.maxZoom);
+
+        const scaleRatio = tr.scale / tr.zoomScale(zoomRef);
+
+        aabb = new Aabb(
+            [aabb.min[0] - (padL + halfScreenPadX) * scaleRatio, aabb.min[1] - (padB + halfScreenPadY) * scaleRatio, aabb.min[2]],
+            [aabb.max[0] + (padR + halfScreenPadX) * scaleRatio, aabb.max[1] + (padT + halfScreenPadY) * scaleRatio, aabb.max[2]]);
+
+        const aabbHalfExtentZ = (aabb.max[2] - aabb.min[2]) * 0.5;
```
Comment (mourner) [inline src/ui/camera.js:795]: Nit: `z` doesn't change:

```suggestion
        const aabbHalfExtentZ = size[2] * 0.5;
```

**Review by mourner — APPROVED** [review]
A few more nits but overall 👍

File: src/ui/camera.js (hunk 14 above)
Comment (karimnaaji) [reply src/ui/camera.js:632]: Yep, good catch

File: src/ui/camera.js (hunk 15 above)
Comment (karimnaaji) [reply src/ui/camera.js]: We might have to leave this one, I just remembered why it was extracted this way:

```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/ui/camera.js:702:45

Cannot call vec3.sub with globeCenter bound to the third parameter because tuple type [1] is incompatible with array
type [2]. [incompatible-call]

     flow-typed/gl-matrix.js
 [2]   2│ type VecType = Array<number> | Float32Array | Float64Array;

     src/geo/transform.js
 [1] 101│     globeCenterInViewSpace: [number, number, number];
```

File: src/ui/camera.js (hunk 9 above)
Comment (karimnaaji) [reply src/ui/camera.js]: Yes `aabb.center` and `origin` may be different, as the AABB may be underneath the surface of the earth, and the `origin` is always at the surface. They may also be misaligned from the camera to center direction, depending on the orientation with pitch and bearing.

**Comment by karimnaaji** [comment]
>I'm seeing some strange behavior with bounds surrounding a pole. The map usually zooms much further out than I would expect. Sometimes, when starting at high zoom and pitch, zoom doesn't change at all.

That's a side from the method we adopted in https://github.com/mapbox/mapbox-gl-js/pull/11951. Minimizing scaling towards the pole has the effect of not resolving the real zoom for a given altitude. Since we re-adjust the scaling to be constant instead of implicitly zooming in but aren't adjusting zoom at the same time, we may resolve a much higher zoom than if we were to zoom towards the pole, compared to before https://github.com/mapbox/mapbox-gl-js/issues/11353.

The current approach towards resolving altitude -> true z in this PR is an approximation: https://github.com/mapbox/mapbox-gl-js/pull/12138/files#diff-a2b3330768267b8c7996b8bf292fcd864e8b768f538436813b0d7e19f7202176R1988, which may introduce a fairly localized issue around the poles/extreme latitudes as you noticed. I think we can improve and put more effort into deriving true zoom if that becomes more of an issue in practice.

A suggested fix towards the original issue of #11353 would be to adjust the zoom dynamically, so that such calculation are always correct, but this had other side effects since styling can depend on zoom. It's not a trivial problem and these are some of the trade-offs that we have to account for. 

>I'm also noticing that the globe will spin around rather than crossing the antimeridian, this seems to be an existing issue with map.easeTo().

This isn't an issue directly related to this PR, and part of `easeTo`.
